### PR TITLE
[9.x] Add `DatabaseTruncation` trait for testing

### PR DESCRIPTION
**Pull request by patrickomeara**
This is a second attempt at #45630

The `DatabaseTruncates` trait is a drop in replacement for `DatabaseMigrations` however it only truncates tables with data instead of dropping and migrating the whole database, this is a more performant approach in most cases.

In this PR I have added support for multiple connections, just like the `RefreshDatabase` trait.

Excluding tables can be done per connection or across all connections depending on what the developer needs.

@emargareten raised the issue of [foreign keys](https://github.com/laravel/framework/pull/45630#discussion_r1071855050) in the last PR. As this PR moves the truncating to the start of the test, I have added a `usesForeignKeyChecks()` which allows the developer to set the check per connection, defaulting to true.

Once this PR is resolved I will PR [this change](https://github.com/patrickomeara/framework/compare/database-truncates-take-2...patrickomeara:framework:without-events) that cleans up the logic around the event dispatcher

### CODE REVIEW

**Comment by patrickomeara** [comment]
Here are some result comparisons against `DatabaseMigrations`

19 Dusk Tests
image

49 Feature Tests
image

> This is for demonstration purposes only. `RefreshDatabase` is a much better choice for Feature Tests.

**Comment by nshiro** [comment]
Nice approach. 😄

But this `DatabaseTruncates` does not seem to work in parallel mode.
`php artisan test --parallel`
I tested on Laravel 9.48.0 / MySQL 8.0.30.

While `DatabaseMigrations` works.

**Comment by taylorotwell** [comment]
@nshiro can you elaborate? error messages?

**Comment by nshiro** [comment]
@taylorotwell
I just made 7 tests for Feature/ExampleTest.php and 7 more tests for Feature/Example2Test.php like below.

(I snipped some)

```php
class ExampleTest extends TestCase
{
    use DatabaseTruncates;
    // use DatabaseMigrations;

    /**
     * A basic test example.
     *
     * @return void
     */
    public function test_the_application_returns_a_successful_response()
    {
        User::factory(20)->create();

        $response = $this->get('/');

        $response->assertStatus(200);
    }

    public function test_the_application_returns_a_successful_response2()
    {
        User::factory(20)->create();

        $response = $this->get('/');

        $response->assertStatus(200);
    }

(snip)

    public function test_the_application_returns_a_successful_response7()
    {
        User::factory(20)->create();

        $response = $this->get('/');

        $response->assertStatus(200);
    }
}
```

```php
class Example2Test extends TestCase
{
    use DatabaseTruncates;
    // use DatabaseMigrations;

    /**
     * A basic test example.
     *
     * @return void
     */
    public function test_the_application_returns_a_successful_response()
    {
        User::factory(20)->create();

        $response = $this->get('/');

        $response->assertStatus(200);
    }

    public function test_the_application_returns_a_successful_response2()
    {
        User::factory(20)->create();

        $response = $this->get('/');

        $response->assertStatus(200);
    }

(snip)

    public function test_the_application_returns_a_successful_response7()
    {
        User::factory(20)->create();

        $response = $this->get('/');

        $response->assertStatus(200);
    }
}
```

When I run tests normally, there are no errors.
But with --parallel option, I get errors below.

(It's too long. I snipped some.)

```
taro ~/proj/temp6> php artisan test --parallel
ParaTest v6.8.1 upon PHPUnit 9.5.28 by Sebastian Bergmann and contributors.

.EEEEEE.EEEEE..                                                   15 / 15 (100%)

Time: 00:01.009, Memory: 20.00 MB

There were 11 errors:

1) Tests\Feature\Example2Test::test_the_application_returns_a_successful_response
Illuminate\Database\QueryException: SQLSTATE[42S02]: Base table or view not found: 1146 Table 'testing.migrations' doesn't exist (SQL: select `migration` from `migrations` order by `batch` asc, `migration` asc)

/home/ftpusers/project/temp6/vendor/laravel/framework/src/Illuminate/Database/Connection.php:760
/home/ftpusers/project/temp6/vendor/laravel/framework/src/Illuminate/Database/Connection.php:720
/home/ftpusers/project/temp6/vendor/laravel/framework/src/Illuminate/Database/Connection.php:405
/home/ftpusers/project/temp6/vendor/laravel/framework/src/Illuminate/Database/Query/Builder.php:2655
/home/ftpusers/project/temp6/vendor/laravel/framework/src/Illuminate/Database/Query/Builder.php:2870
/home/ftpusers/project/temp6/vendor/laravel/framework/src/Illuminate/Database/Query/Builder.php:3180
/home/ftpusers/project/temp6/vendor/laravel/framework/src/Illuminate/Database/Query/Builder.php:2866
/home/ftpusers/project/temp6/vendor/laravel/framework/src/Illuminate/Database/Migrations/DatabaseMigrationRepository.php:53
/home/ftpusers/project/temp6/vendor/laravel/framework/src/Illuminate/Database/Migrations/Migrator.php:692
/home/ftpusers/project/temp6/vendor/laravel/framework/src/Illuminate/Database/Console/Migrations/MigrateCommand.php:126
/home/ftpusers/project/temp6/vendor/laravel/framework/src/Illuminate/Database/Console/Migrations/MigrateCommand.php:82
/home/ftpusers/project/temp6/vendor/laravel/framework/src/Illuminate/Database/Migrations/Migrator.php:615
/home/ftpusers/project/temp6/vendor/laravel/framework/src/Illuminate/Database/Console/Migrations/MigrateCommand.php:81
/home/ftpusers/project/temp6/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php:36
/home/ftpusers/project/temp6/vendor/laravel/framework/src/Illuminate/Container/Util.php:41
/home/ftpusers/project/temp6/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php:93
/home/ftpusers/project/temp6/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php:35
/home/ftpusers/project/temp6/vendor/laravel/framework/src/Illuminate/Container/Container.php:663
/home/ftpusers/project/temp6/vendor/laravel/framework/src/Illuminate/Console/Command.php:182
/home/ftpusers/project/temp6/vendor/symfony/console/Command/Command.php:312
/home/ftpusers/project/temp6/vendor/laravel/framework/src/Illuminate/Console/Command.php:151
/home/ftpusers/project/temp6/vendor/laravel/framework/src/Illuminate/Console/Concerns/CallsCommands.php:67
/home/ftpusers/project/temp6/vendor/laravel/framework/src/Illuminate/Console/Concerns/CallsCommands.php:28
/home/ftpusers/project/temp6/vendor/laravel/framework/src/Illuminate/Database/Console/Migrations/FreshCommand.php:53
/home/ftpusers/project/temp6/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php:36
/home/ftpusers/project/temp6/vendor/laravel/framework/src/Illuminate/Container/Util.php:41
/home/ftpusers/project/temp6/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php:93
/home/ftpusers/project/temp6/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php:35
/home/ftpusers/project/temp6/vendor/laravel/framework/src/Illuminate/Container/Container.php:663
/home/ftpusers/project/temp6/vendor/laravel/framework/src/Illuminate/Console/Command.php:182
/home/ftpusers/project/temp6/vendor/symfony/console/Command/Command.php:312
/home/ftpusers/project/temp6/vendor/laravel/framework/src/Illuminate/Console/Command.php:151
/home/ftpusers/project/temp6/vendor/symfony/console/Application.php:1022
/home/ftpusers/project/temp6/vendor/symfony/console/Application.php:314
/home/ftpusers/project/temp6/vendor/symfony/console/Application.php:168
/home/ftpusers/project/temp6/vendor/laravel/framework/src/Illuminate/Console/Application.php:102
/home/ftpusers/project/temp6/vendor/laravel/framework/src/Illuminate/Console/Application.php:193
/home/ftpusers/project/temp6/vendor/laravel/framework/src/Illuminate/Foundation/Console/Kernel.php:336
/home/ftpusers/project/temp6/vendor/laravel/framework/src/Illuminate/Testing/PendingCommand.php:296
/home/ftpusers/project/temp6/vendor/laravel/framework/src/Illuminate/Testing/PendingCommand.php:481
/home/ftpusers/project/temp6/vendor/laravel/framework/src/Illuminate/Foundation/Testing/Concerns/InteractsWithConsole.php:80
/home/ftpusers/project/temp6/vendor/laravel/framework/src/Illuminate/Foundation/Testing/DatabaseTruncates.php:19
/home/ftpusers/project/temp6/vendor/laravel/framework/src/Illuminate/Foundation/Testing/TestCase.php:135
/home/ftpusers/project/temp6/vendor/laravel/framework/src/Illuminate/Foundation/Testing/TestCase.php:96

2) Tests\Feature\Example2Test::test_the_application_returns_a_successful_response2
Illuminate\Database\QueryException: SQLSTATE[42S01]: Base table or view already exists: 1050 Table 'migrations' already exists (SQL: create table `migrations` (`id` int unsigned not null auto_increment primary key, `migration` varchar(255) not null, `batch` int not null) default character set utf8mb4 collate 'utf8mb4_unicode_ci')

/home/ftpusers/project/temp6/vendor/laravel/framework/src/Illuminate/Database/Connection.php:760
/home/ftpusers/project/temp6/vendor/laravel/framework/src/Illuminate/Database/Connection.php:720
/home/ftpusers/project/temp6/vendor/laravel/framework/src/Illuminate/Database/Connection.php:534
(snip)
/home/ftpusers/project/temp6/vendor/laravel/framework/src/Illuminate/Foundation/Testing/TestCase.php:135
/home/ftpusers/project/temp6/vendor/laravel/framework/src/Illuminate/Foundation/Testing/TestCase.php:96

3) Tests\Feature\Example2Test::test_the_application_returns_a_successful_response3
Illuminate\Database\QueryException: SQLSTATE[42S01]: Base table or view already exists: 1050 Table 'users' already exists (SQL: create table `users` (`id` bigint unsigned not null auto_increment primary key, `name` varchar(255) not null, `email` varchar(255) not null, `email_verified_at` timestamp null, `password` varchar(255) not null, `remember_token` varchar(100) null, `created_at` timestamp null, `updated_at` timestamp null) default character set utf8mb4 collate 'utf8mb4_unicode_ci')

/home/ftpusers/project/temp6/vendor/laravel/framework/src/Illuminate/Database/Connection.php:760
/home/ftpusers/project/temp6/vendor/laravel/framework/src/Illuminate/Database/Connection.php:720
/home/ftpusers/project/temp6/vendor/laravel/framework/src/Illuminate/Database/Connection.php:534
(snip)
/home/ftpusers/project/temp6/vendor/laravel/framework/src/Illuminate/Foundation/Testing/TestCase.php:135
/home/ftpusers/project/temp6/vendor/laravel/framework/src/Illuminate/Foundation/Testing/TestCase.php:96

4) Tests\Feature\Example2Test::test_the_application_returns_a_successful_response5
Illuminate\Database\QueryException: SQLSTATE[42S02]: Base table or view not found: 1146 Table 'testing.users' doesn't exist (SQL: insert into `users` (`name`, `email`, `email_verified_at`, `password`, `remember_token`, `updated_at`, `created_at`) values (Miss Santina Herzog Sr., laurel76@example.net, 2023-01-27 12:39:21, $2y$10$92IXUNpkjO0rOQ5byMi.Ye4oKoEa3Ro9llC/.og/at2.uheWG/igi, gdVLLI3P7j, 2023-01-27 12:39:21, 2023-01-27 12:39:21))

/home/ftpusers/project/temp6/vendor/laravel/framework/src/Illuminate/Database/Connection.php:760
/home/ftpusers/project/temp6/vendor/laravel/framework/src/Illuminate/Database/Connection.php:720
/home/ftpusers/project/temp6/vendor/laravel/framework/src/Illuminate/Database/Connection.php:534
(snip)
/home/ftpusers/project/temp6/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Factories/Factory.php:325
/home/ftpusers/project/temp6/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Factories/Factory.php:283
/home/ftpusers/project/temp6/tests/Feature/Example2Test.php:60

5) Tests\Feature\Example2Test::test_the_application_returns_a_successful_response6
Illuminate\Database\QueryException: SQLSTATE[42S02]: Base table or view not found: 1146 Table 'testing.password_resets' doesn't exist (SQL: select exists(select * from `password_resets`) as `exists`)

/home/ftpusers/project/temp6/vendor/laravel/framework/src/Illuminate/Database/Connection.php:760
/home/ftpusers/project/temp6/vendor/laravel/framework/src/Illuminate/Database/Connection.php:720
/home/ftpusers/project/temp6/vendor/laravel/framework/src/Illuminate/Database/Connection.php:405
(snip)
/home/ftpusers/project/temp6/vendor/laravel/framework/src/Illuminate/Foundation/Testing/TestCase.php:135
/home/ftpusers/project/temp6/vendor/laravel/framework/src/Illuminate/Foundation/Testing/TestCase.php:96

6) Tests\Feature\Example2Test::test_the_application_returns_a_successful_response7
Illuminate\Database\QueryException: SQLSTATE[42S02]: Base table or view not found: 1146 Table 'testing.password_resets' doesn't exist (SQL: select exists(select * from `password_resets`) as `exists`)

/home/ftpusers/project/temp6/vendor/laravel/framework/src/Illuminate/Database/Connection.php:760
/home/ftpusers/project/temp6/vendor/laravel/framework/src/Illuminate/Database/Connection.php:720
/home/ftpusers/project/temp6/vendor/laravel/framework/src/Illuminate/Database/Connection.php:405
(snip)
/home/ftpusers/project/temp6/vendor/laravel/framework/src/Illuminate/Foundation/Testing/TestCase.php:135
/home/ftpusers/project/temp6/vendor/laravel/framework/src/Illuminate/Foundation/Testing/TestCase.php:96

(snip)

11) Tests\Feature\ExampleTest::test_the_application_returns_a_successful_response5
Illuminate\Database\QueryException: SQLSTATE[42S01]: Base table or view already exists: 1050 Table 'personal_access_tokens' already exists (SQL: create table `personal_access_tokens` (`id` bigint unsigned not null auto_increment primary key, `tokenable_type` varchar(255) not null, `tokenable_id` bigint unsigned not null, `name` varchar(255) not null, `token` varchar(64) not null, `last_used_at` timestamp null, `expires_at` timestamp null, `created_at` timestamp null, `updated_at` timestamp null) default character set utf8mb4 collate 'utf8mb4_unicode_ci')

/home/ftpusers/project/temp6/vendor/laravel/framework/src/Illuminate/Database/Connection.php:760
/home/ftpusers/project/temp6/vendor/laravel/framework/src/Illuminate/Database/Connection.php:720
/home/ftpusers/project/temp6/vendor/laravel/framework/src/Illuminate/Database/Connection.php:534
(snip)
/home/ftpusers/project/temp6/vendor/laravel/framework/src/Illuminate/Foundation/Testing/DatabaseTruncates.php:19
/home/ftpusers/project/temp6/vendor/laravel/framework/src/Illuminate/Foundation/Testing/TestCase.php:135
/home/ftpusers/project/temp6/vendor/laravel/framework/src/Illuminate/Foundation/Testing/TestCase.php:96

FAILURES!
Tests: 15, Assertions: 4, Errors: 11.
```



When I use DatabaseMigrations, there are no problems at all.

**Comment by nshiro** [comment]
I also noticed I needed `doctrine/dbal` to use DatabaseTruncates.
`composer require doctrine/dbal`

**Comment by patrickomeara** [comment]
Thanks for your feedback @nshiro 

Are you able to create a repository with the failing tests?

I have created one [here](https://github.com/patrickomeara/truncates) based on what you mentioned and all tests pass with and without parallel.

image

Here is the diff from a fresh install
https://github.com/patrickomeara/truncates/compare/194771c5830a16c8f06f2eef34e7c28bf6a4c26d...main

Yes `doctrine/dbal` is a dev requirement to get the table names for all database grammar's. This should be mentioned in the [Dusk docs](https://laravel.com/docs/master/dusk#migrations) if this PR is accepted. As it is [here](https://laravel.com/docs/master/migrations#prerequisites).

**Comment by nshiro** [comment]
@patrickomeara
I apologise to you. It was all my fault. 🙇‍♂️🙇‍♂️
I forgot to copy and paste the last bit of your code.
Now the test works in parallel mode too.
Great work !

**Comment by patrickomeara** [comment]
@nshiro Awesome. Thanks for testing this PR out.

**Comment by taylorotwell** [comment]
Thanks - could you PR some docs to laravel/docs where you think this is relevant?

File: src/Illuminate/Foundation/Testing/DatabaseTruncation.php [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+
+            return;
+        }
+
+        // Always clear any test data on subsequent runs...
+        $this->truncateTablesForAllConnections();
+
+        if ($seeder = $this->seeder()) {
+            // Use a specific seeder class...
+            $this->artisan('db:seed', ['--class' => $seeder]);
```
Comment (antonkomarev) [inline src/Illuminate/Foundation/Testing/DatabaseTruncation.php:43]: Maybe it will be better to use FQCN instead of the command name?

**Comment by patrickomeara** [comment]
> Thanks - could you PR some docs to laravel/docs where you think this is relevant?

@taylorotwell Yes, will do this ASAP.

**Pull request merged**